KAFKA-15818: ensure leave group on max poll interval

philipnee opened the pull request:
Currently, poll interval is not being respected during consumer#poll.  When the user stops polling the consumer, we should assume either the consumer is too slow to respond or is already dead.  In either case, we should let the group coordinator kick the member out of the group and reassign its partition after the rebalance timeout expires.  

If the consumer comes back alive, we should send a heartbeat and the member will be fenced and rejoin.  (and the partitions will be revoked).

This is the same behavior as the current implementation.

philipnee committed:
Ensure poll timer stops heartbeat

refactor

updates

Revert "hacked to work"

This reverts commit c43382bde611df5275a0daf1f6fa3641609a91f8.

hacked to work

Update HeartbeatRequestManager.java

updates

leavegroup on polltimer expiration

Basically poll interval impl

wip

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:
```diff
@@ -169,6 +174,19 @@ public NetworkClientDelegate.PollResult poll(long currentTimeMs) {
+                "the poll loop is spending too much time processing messages. You can address this " +
+                "either by increasing max.poll.interval.ms or by reducing the maximum size of batches " +
+                "returned in poll() with max.poll.records.");
+            membershipManager.leaveGroup().whenComplete((ignored, exception) -> {
```
Not sure if kip-848 changes this but
```
// Starting from 2.3, only dynamic members will send LeaveGroupRequest to the broker,
        // consumer with valid group.instance.id is viewed as static member that never sends LeaveGroup,
        // and the membership expiration is only controlled by session timeout.
```

philipnee replied:
What is the exact to do in KIP-848 if the poll timer expires? The current implementation sends a leavegroup request directly without revoking partitions.  If we want to do that, we need to modify the state transition a bit.

philipnee commented:
In the current implementation, when poll timer expires, the consumer sends a leavegroup request w/o revoking the partitions. This is obviously a bit different to the current state transition, so I'm not sure what's the right thing to do.

philipnee committed:
ensure leave group is sent upon polltimer expires.  Joining when the consumer is polled again.

Update HeartbeatRequestManager.java

clean up

philipnee commented:
hi @lianetm @lucasbru can I ask for a review for this PR? Modification was made based on our discussion.

lucasbru commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:
```diff
@@ -121,11 +127,13 @@ public HeartbeatRequestManager(
         this.heartbeatState = new HeartbeatState(subscriptions, membershipManager, rebalanceTimeoutMs);
         this.heartbeatRequestState = new HeartbeatRequestState(logContext, time, 0, retryBackoffMs,
             retryBackoffMaxMs, rebalanceTimeoutMs);
+        this.pollTimer = time.timer(rebalanceTimeoutMs);
```
rebalanceTimeoutMs == maxPollIntervalMs ? maybe rename in that case.

cadonna replied:
Yes, please, rename.

lucasbru commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:
```diff
@@ -198,6 +219,17 @@ public long maximumTimeToWait(long currentTimeMs) {
+    /**
+     * When consumer polls, we need to reset the pollTimer.  If member is already leaving the group
+     */
+    public void ack() {
```
Is the timer up to date at this point? Otherwise, `deadlineMs` will be set incorrectly. Maybe pass in the current time into this function and update the timer before setting the timeout. Like we do for auto-commit.

Also, better name please. `resetPollTimer` ?

cadonna replied:
+1

philipnee replied:
The code is updated - but the pollTimer is updated on a regular cadence by the HeartbeatRequestManager.poll(). this func is used to check if the timer has been expired.  If it was, then we reset the timer everytime we poll the consumer.

lucasbru commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:
```diff
@@ -198,6 +219,17 @@ public long maximumTimeToWait(long currentTimeMs) {
+        pollTimer.reset(rebalanceTimeoutMs);
+    }
+
+    Timer pollTimer() {
```
// visible for testing

cadonna replied:
See my comment above about directly passing the poll timer into the constructor.

lucasbru replied:
I was assuming that after passing the poll timer from the outside, we don't need to make this visible for testing anymore

lucasbru commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -428,9 +428,9 @@ private void updateSubscription(Collection<TopicPartition> assignedPartitions,
      * Transition to the {@link MemberState#JOINING} state, indicating that the member will
      * try to join the group on the next heartbeat request. This is expected to be invoked when
      * the user calls the subscribe API, or when the member wants to rejoin after getting fenced.
-     * Visible for testing.
```
I think it's still visible for testing only

lucasbru commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -399,14 +400,45 @@ public void testHeartbeatState() {
 
+    @Test
+    public void testEnsureLeaveGroupWhenPollTimerExpires() {
+        membershipManager.transitionToJoining();
```
You could consider using `membershipManager.transitionToFenced(); // And indirect way of moving to JOINING state` instead of changing the visibility of prod code

philipnee replied:
it is not possible to transition to fenced from unsubscribed

lianetm replied:
I know we've been all writing this kind of unit tests that go into the integration test land, but if we want to start correcting/simplifying that, this unit test could be simplified. As it is for the HB manager, it should just `verify` the call to the membershipManager func when the timer expires (and then another test in the MembershipMgr for that onStale func, that is needed anyways I would say)

philipnee replied:
you are 100% correct.  I submit a patch for this specific test.  Thanks for catching this.

lucasbru reviewed (commented):
Thanks for the PR! I left a few comments

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:
```diff
@@ -166,11 +175,23 @@ public HeartbeatRequestManager(
+                "the poll loop is spending too much time processing messages. You can address this " +
+                "either by increasing max.poll.interval.ms or by reducing the maximum size of batches " +
+                "returned in poll() with max.poll.records.");
+            return PollResult.EMPTY;
```
I'm thinking what about instead of returning the empty here, we transition the member to "leaving" (`transitionToSendingLeaveGroup`), and that's it. On the following lines that should make the shouldHbNow true, we would send the last HB and transition to unsubscribed (effectively leaving without triggering the callbacks, which is what the old logic does)

cadonna replied:
What @lianetm writes makes sense to me. Looking at the [code of the legacy consumer](https://github.com/apache/kafka/blob/bbcf40ad0dc7739d803244bbbfbaa4598850344b/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java#L1491) the consumer should proactively leave the group and not stop heartbeating and wait until it is kicked out of the group.
However, if `group.instance.id` is set, it should not leave the group and stop heartbeating.

cadonna commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:
```diff
@@ -103,6 +103,12 @@ public class HeartbeatRequestManager implements RequestManager {
     private final BackgroundEventHandler backgroundEventHandler;
 
+    /**
+     * Timer for tracking the time since the last consumer poll.  If the timer expires, the consumer will stop
+     * sending heartbeat until the next poll.
+     */
```
Isn't this only true for when `group.instance.id` is set? That is my interpretation of the following doc: https://kafka.apache.org/documentation/#consumerconfigs_max.poll.interval.ms

philipnee replied:
well, i'm not sure actually.  According to KIP-848: `Static membership, introduced in KIP-345, is still supported by this new rebalance protocol. When a member wants to leave temporary – e.g. while being bounced – it should send an heartbeat with a member epoch equals to -2. This signals to the group coordinator that the member left but will rejoin within the session timeout. When the member rejoins with the same instance ID, the group coordinator replaces the old member by the new member and gives back its current assignment.`

philipnee replied:
I think you meant : only when `group.instance.id` is not set, i.e. dynamic member?

This is the current implementation
```
 // Starting from 2.3, only dynamic members will send LeaveGroupRequest to the broker,
        // consumer with valid group.instance.id is viewed as static member that never sends LeaveGroup,
        // and the membership expiration is only controlled by session timeout.
        if (isDynamicMember() && !coordinatorUnknown() &&
            state != MemberState.UNJOINED && generation.hasMemberId()) {
            // this is a minimal effort attempt to leave the group. we do not
            // attempt any resending if the request fails or times out.
            log.info("Member {} sending LeaveGroup request to coordinator {} due to {}",
                generation.memberId, coordinator, leaveReason);
            LeaveGroupRequest.Builder request = new LeaveGroupRequest.Builder(
                rebalanceConfig.groupId,
                Collections.singletonList(new MemberIdentity().setMemberId(generation.memberId).setReason(JoinGroupRequest.maybeTruncateReason(leaveReason)))
            );

            future = client.send(coordinator, request).compose(new LeaveGroupResponseHandler(generation));
            client.pollNoWakeup();
        }
```

lianetm replied:
My take is that we should always send the leave group when the poll timer expires (group.instance.id null or not). Dynamic members should send -1 and static -2 as described in the KIP. The [documentation of the max.poll.interval](https://kafka.apache.org/documentation/#consumerconfigs_max.poll.interval.ms) is just describing what will happen on the broker side (if static member leaving, broker won't re-assign the partitions until the session timeout expires. If dynamic member, it will re-assign them right away).

cadonna replied:
Oh, I was not aware of the static membership handling in KIP-848. I should have also looked into that instead of only looking into the legacy consumer. Thanks for the clarification!

Yes, I agree that consumers that use group management should always send a leave group as KIP-848 describes.

So the comment is wrong. I would propose to just remove the whole comment and give the timer a descriptive name. In general I find inline comments problematic because -- at some point -- they start to lie because they are never properly maintained. Instead I prefer to use meaningful and descriptive names and to extract code that need a comment to methods that either have a meaningful name or a comment on the method, preferring the meaningful name. What I just wrote applies in most cases but not in all, but it definitely applies here.

cadonna commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:
```diff
@@ -121,11 +127,13 @@ public HeartbeatRequestManager(
     // Visible for testing
     HeartbeatRequestManager(
         final LogContext logContext,
+        final Time time,
```
Why not passing in directly the poll timer instead of the mock time that is only used to create the poll timer. By passing in the poll timer, you can also remove method `pollTimer()` since that is only needed for testing.

cadonna commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -399,14 +400,45 @@ public void testHeartbeatState() {
+        assertFalse(heartbeatRequestManager.pollTimer().isExpired());
+        // Expires the poll timer, and ensure heartbeat is not sent
+        time.sleep(DEFAULT_MAX_POLL_INTERVAL_MS);
+        assertNoHeartbeat(heartbeatRequestManager);
```
As I wrote before, I think the consumer should leave the group.

cadonna reviewed (commented):
Thanks for the PR @philipnee !

Here my feedback.

philipnee committed:
PR comments

philipnee committed:
Update HeartbeatRequestManager.java

philipnee commented:
Hi @lianetm @cadonna @lucasbru - Addressed your comments.  I've left a question around sending leave group for static member.  From reading the KIP, seems like we should do it.  I think we need further clarification.

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -399,14 +400,51 @@ public void testHeartbeatState() {
+        assertEquals(1, pollResult.unsentRequests.size());
+        ConsumerGroupHeartbeatRequestData data = (ConsumerGroupHeartbeatRequestData) pollResult.unsentRequests.get(0).requestBuilder().build().data();
+        assertEquals(ConsumerGroupHeartbeatRequest.LEAVE_GROUP_MEMBER_EPOCH, data.memberEpoch());
+        assertEquals(MemberState.UNSUBSCRIBED, membershipManager.state());
```
Is this really the right end state when the poll timer expires? This means that the user will have to call `subscribe` to join the group again. Is that what the old coordinator requires? (I expected that the old code just requires the consumer to be polled again to join the group. If my understanding is right then we're missing logic after sending the last HB to leave: on timer expiration we should send the last HB and transitionToJoining so that the member re-joins the group on the next poll)

lianetm commented:
Thanks for the changes @philipnee, left answers for the static membership and some comments.

philipnee committed:
implement STALED

philipnee commented:
Hi @lianetm - Thanks for the time reviewing the PR. Instead of using boolean flags, I added a _STALED_ state to address your concern.  I think we do need this state to stay consistent with the state logic, i.e., I agree that unsubscribed is probably not the right state to be in as it hints to the user that it needs to subscribe.  Let me know what do you think.

For the static member - I think the right thing to do is to let the membershipManager to decide the right epoch to send.

lucasbru commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManager.java:
```diff
@@ -137,4 +142,15 @@ public interface MembershipManager {
+    void transitionToJoining();
+
+    /**
+     * When the user stops polling the consumer, the member will be transitioned to LEAVING without revoking the
```
How does the comment about `LEAVING` relate to this method? Maybe needs to be updated?

philipnee committed:
Addressing PR comments

Update HeartbeatRequestManager.java

clean up

PR comments

implement STALED

Update HeartbeatRequestManager.java

PR comments

Ensure poll timer stops heartbeat

refactor

updates

Revert "hacked to work"

This reverts commit c43382bde611df5275a0daf1f6fa3641609a91f8.

hacked to work

Update HeartbeatRequestManager.java

updates

leavegroup on polltimer expiration

Basically poll interval impl

wip

resolving conflict

PR comments

update a comment

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:
```diff
@@ -201,13 +224,35 @@ public long maximumTimeToWait(long currentTimeMs) {
+                // rejoin on the next poll regardless the server has responded to the heartbeat.
+                if (!ignoreResponse) {
+                    membershipManager.onHeartbeatResponseReceived(((ConsumerGroupHeartbeatResponse) response.responseBody()).data());
+                    maybeSendGroupMetadataUpdateEvent();
```
@cadonna - can you comment on this change?

cadonna replied:
If the heartbeat request manager ignores the response, why does it call `onResponse()`? Does the heartbeat request manager continue to send heartbeats although the consumer left the group? If yes, what is the reason for it? I would have expected that the consumer leaves the group and stops heartbeating. Then at the next call to `Consumer#poll()` it should join the group again.

Another question, how does the consumer avoid that offsets are committed (or any other group related request) when the poll timer expired and the consumer is in the staled state? Probably the broker will reject the commit, but actually the consumer proactively left the group, so it knows that it is not part of the group anymore and it could just not send the offset commit.

cadonna replied:
@philipnee Any thoughts about my comment?

philipnee replied:
For calling onResponse: We need to try our best to handle the error if there is one.   However, when leaving a group, we don't need to process the `onHeartbeatResponseReceived` because we don't need to reconcile.

After leaving the group - The request manager should not keep heart beating.  For this specific PR, when the poll timer expires, it will return an empty response when the request manager gets polled.  See the first if block in poll().  And you are right, when the user poll the consumer, it resets the timer, and on the next request manager poll (they happen asynchronously), it should send a heartbeat to rejoin.

i think if we commit after leaving the group - I think the commit will get rejected with an error.

lianetm replied:
To complete the pic regarding the commits, if a commit was sent out right before leaving, it will include memberId and epoch that are not valid anymore, so the broker will reject as @philipnee mentioned, sending an UNKNOWN_MEMBER_ID, and the client will fail the commit with that unrecoverable error. That being said, following commit requests issued after the consumer left the group will succeed because they will be sent without member ID or epoch (noticing that the member is not part of the group anymore). That logic is in the commitRequestManager but sharing here to connect the dots.

philipnee replied:
thanks for the clarification @lianetm

cadonna replied:
OK, thank you for the clarification!
Sorry, but I still have two comments:

1. Why to we still need to call `onResponse()`? As far as I understand, all what we do there is not strictly needed because we will do the same when we restart heartbeating in the next poll, won't we? Is the error handling in this case so important? Wouldn't it be equally fine to not handle errors until we start heartbeating? I am asking because the code and probably also the reasoning about the code would become simpler if we just completely ignored the response. I also see that it is not a big deal doing it this or the other way. So, if you do not want to change it, it is fine with me.
However, I just realized that with this change we will call `membershipManager.onHeartbeatResponseReceived()` and `maybeSendGroupMetadataUpdateEvent()` in the error case, won't we? Before the change these two calls are only done without an error. Is this acceptable?

2. I understand that the commit gets rejected by the brokers. My question was whether we should avoid sending a commit request in the first place, since the consumer actually knows that it is not part of a group anymore. 

A last question:

> That being said, following commit requests issued after the consumer left the group will succeed because they will be sent without member ID or epoch (noticing that the member is not part of the group anymore).

I am not sure I understand this. Why does the commit succeed?

philipnee committed:
Refactor tests to avoid spying

philipnee committed:
Update MembershipManagerImplTest.java

lucasbru reviewed (commented):
modulo the discussion about group metadata, this looks good to me

philipnee commented:
There's quite a bit of failing tests, so i rebased to try to resolve them.

lucasbru commented:
There is a suspicious test failure: 

https://ge.apache.org/scans/tests?search.relativeStartTime=P28D&search.rootProjectNames=kafka&search.timeZoneId=Europe%2FVienna&tests.container=kafka.api.PlaintextConsumerTest&tests.test=testRemoteAssignorRange(String%2C%20String)%5B1%5D

I doesn't seem to fail on master at all, but it fails consistently for this PR. It failed on one other PR once. We probably have to look at this

lucasbru reviewed (approved):
LGTM, however we still need to do some minor changes and if possible testing before we merge.

For the failing test, can you please (in this PR) increase the MAX_POLL_INTERVAL_MS to 30 seconds or so? In most integration tests its set to 6s which doesn't seem to be enough with the new consumer. We should create a ticket to investigate why it's so slow, but I don't think there is something wrong with the PR. 

More generally, can you please run the integration tests with this PR, since we want to enable them on trunk, and they are currently not run by CI (except for some that Andrew did not disable).

Can you also enable the `testMaxPollIntervalMs` test and some of the other integration tests around max_poll_interval_ms if they are stable? Could be in a separate PR though, so we can merge this one.

@lianetm / @cadonna any more comments on this?

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -587,6 +599,18 @@ public boolean shouldSkipHeartbeat() {
+     */
+    @Override
+    public void transitionToStaled() {
+        transitionTo(MemberState.PREPARE_LEAVING);
```
Looks to me this is not really needed here? We use this state when we are taking some action before sending the last HB (mainly callbacks), but in this case there is no async operation to do, just send the last HB right? which is explicitly done in the HB mgr when checking that the poll timer expired

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -587,6 +599,18 @@ public boolean shouldSkipHeartbeat() {
+    public void transitionToStaled() {
+        transitionTo(MemberState.PREPARE_LEAVING);
+        memberEpoch = ConsumerGroupHeartbeatRequest.LEAVE_GROUP_MEMBER_EPOCH;
+        transitionTo(MemberState.LEAVING);
```
Similar, do we need this? We use this LEAVING state to signal the HB manager that it should send a HB now, but in this staled member case, that last HB is sent based on the check that the poll timer expired (not based on any state that should HB now)

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -587,6 +599,18 @@ public boolean shouldSkipHeartbeat() {
+        memberEpoch = ConsumerGroupHeartbeatRequest.LEAVE_GROUP_MEMBER_EPOCH;
+        transitionTo(MemberState.LEAVING);
+        leaveGroupInProgress = Optional.of(CompletableFuture.completedFuture(null));
+        transitionTo(MemberState.STALED);
```
We also need to make sure that when going down this path, we clear the current assignment, so the member actually rejoins with its subscription, but ready to take on whatever new partitions the broker assigns.

philipnee replied:
thanks that makes a lot of sense.

philipnee replied:
I think transitionToJoining, which takes place when sending out the heartbeat, already clears the current assignment.

philipnee replied:
nvm - i was confused by the pending assignments.  I think we need to clear the current assignment.

philipnee committed:
pr comment

lianetm commented:
Thanks for the changes @philipnee , LGTM. 

I would only +1 on @lucasbru suggestion of locally running the existing `PlainTextConsumer` integration test on this PR, as it is currently disabled in the CI. Also trying to get some of the max poll related ones to run (OK if in different PR, as I expect it will require some changes due to the listeners they use that we don't support yet).

lucasbru commented:
RemoteRangeAssignorTest is still failing for this one

cadonna commented:
> @lianetm / @cadonna any more comments on this?

@lucasbru I did not get any answer to the following comment: https://github.com/apache/kafka/pull/14873#discussion_r1418729793

philipnee commented:
hi @cadonna - Sorry about unintentionally ignoring your question.  I've just responded to it.

philipnee commented:
Hi @lianetm and @lucasbru - Per your request, I've just ran the tests using 

```
Arguments.of("kraft+kip848", "consumer"))
```

All tests passed except `testRemoteAssignorRange`.  Per lucas suggestion, I extended the max.poll.interval.ms to 30s, the test passed.

```
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testCoordinatorFailover(String, String) > testCoordinatorFailover(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testCoordinatorFailover(String, String) > testCoordinatorFailover(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testCoordinatorFailover(String, String) > testCoordinatorFailover(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testClusterResourceListener(String, String) > testClusterResourceListener(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testSimpleConsumption(String, String) > testSimpleConsumption(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPartitionsForAutoCreate(String, String) > testPartitionsForAutoCreate(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPartitionsForAutoCreate(String, String) > testPartitionsForAutoCreate(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPartitionsForAutoCreate(String, String) > testPartitionsForAutoCreate(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testShrinkingTopicSubscriptions(String, String) > testShrinkingTopicSubscriptions(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testShrinkingTopicSubscriptions(String, String) > testShrinkingTopicSubscriptions(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testShrinkingTopicSubscriptions(String, String) > testShrinkingTopicSubscriptions(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMaxPollIntervalMs(String, String) > testMaxPollIntervalMs(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMaxPollIntervalMs(String, String) > testMaxPollIntervalMs(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMaxPollIntervalMs(String, String) > testMaxPollIntervalMs(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testAssignAndConsumeFromCommittedOffsets(String, String) > testAssignAndConsumeFromCommittedOffsets(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testSubsequentPatternSubscription(String, String) > testSubsequentPatternSubscription(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testSubsequentPatternSubscription(String, String) > testSubsequentPatternSubscription(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testSubsequentPatternSubscription(String, String) > testSubsequentPatternSubscription(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPerPartitionLeadMetricsCleanUpWithAssign(String, String) > testPerPartitionLeadMetricsCleanUpWithAssign(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testNullGroupIdNotSupportedIfCommitting(String, String) > testNullGroupIdNotSupportedIfCommitting(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testEmptyGroupIdNotSupported(String, String) > testEmptyGroupIdNotSupported(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testAssignAndCommitAsyncNotCommitted(String, String) > testAssignAndCommitAsyncNotCommitted(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testConsumeMessagesWithCreateTime(String, String) > testConsumeMessagesWithCreateTime(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testAsyncCommit(String, String) > testAsyncCommit(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testLowMaxFetchSizeForRequestAndPartition(String, String) > testLowMaxFetchSizeForRequestAndPartition(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMultiConsumerSessionTimeoutOnStopPolling(String, String) > testMultiConsumerSessionTimeoutOnStopPolling(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMultiConsumerSessionTimeoutOnStopPolling(String, String) > testMultiConsumerSessionTimeoutOnStopPolling(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMultiConsumerSessionTimeoutOnStopPolling(String, String) > testMultiConsumerSessionTimeoutOnStopPolling(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMaxPollIntervalMsDelayInRevocation(String, String) > testMaxPollIntervalMsDelayInRevocation(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMaxPollIntervalMsDelayInRevocation(String, String) > testMaxPollIntervalMsDelayInRevocation(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMaxPollIntervalMsDelayInRevocation(String, String) > testMaxPollIntervalMsDelayInRevocation(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPerPartitionLagMetricsCleanUpWithAssign(String, String) > testPerPartitionLagMetricsCleanUpWithAssign(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPartitionsForInvalidTopic(String, String) > testPartitionsForInvalidTopic(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPartitionsForInvalidTopic(String, String) > testPartitionsForInvalidTopic(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPartitionsForInvalidTopic(String, String) > testPartitionsForInvalidTopic(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPauseStateNotPreservedByRebalance(String, String) > testPauseStateNotPreservedByRebalance(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testFetchHonoursFetchSizeIfLargeRecordNotFirst(String, String) > testFetchHonoursFetchSizeIfLargeRecordNotFirst(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testAssignAndRetrievingCommittedOffsetsMultipleTimes(String, String) > testAssignAndRetrievingCommittedOffsetsMultipleTimes(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testSeek(String, String) > testSeek(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testConsumingWithNullGroupId(String, String) > testConsumingWithNullGroupId(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPositionAndCommit(String, String) > testPositionAndCommit(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testFetchRecordLargerThanMaxPartitionFetchBytes(String, String) > testFetchRecordLargerThanMaxPartitionFetchBytes(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testRebalanceAndRejoin(String, String) > [1] assignmentStrategy=org.apache.kafka.clients.consumer.CooperativeStickyAssignor, quorum=zk PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testRebalanceAndRejoin(String, String) > [2] assignmentStrategy=org.apache.kafka.clients.consumer.RangeAssignor, quorum=zk PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testRebalanceAndRejoin(String, String) > [3] assignmentStrategy=org.apache.kafka.clients.consumer.CooperativeStickyAssignor, quorum=kraft PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testRebalanceAndRejoin(String, String) > [4] assignmentStrategy=org.apache.kafka.clients.consumer.RangeAssignor, quorum=kraft PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testUnsubscribeTopic(String, String) > testUnsubscribeTopic(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testUnsubscribeTopic(String, String) > testUnsubscribeTopic(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testUnsubscribeTopic(String, String) > testUnsubscribeTopic(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMultiConsumerSessionTimeoutOnClose(String, String) > testMultiConsumerSessionTimeoutOnClose(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMultiConsumerSessionTimeoutOnClose(String, String) > testMultiConsumerSessionTimeoutOnClose(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMultiConsumerSessionTimeoutOnClose(String, String) > testMultiConsumerSessionTimeoutOnClose(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testRemoteAssignorRange(String, String) > testRemoteAssignorRange(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMultiConsumerStickyAssignor(String, String) > testMultiConsumerStickyAssignor(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMultiConsumerStickyAssignor(String, String) > testMultiConsumerStickyAssignor(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMultiConsumerStickyAssignor(String, String) > testMultiConsumerStickyAssignor(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testAssignAndConsumeSkippingPosition(String, String) > testAssignAndConsumeSkippingPosition(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testFetchRecordLargerThanFetchMaxBytes(String, String) > testFetchRecordLargerThanFetchMaxBytes(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testAssignAndConsumeWithLeaderChangeValidatingPositions(String, String) > testAssignAndConsumeWithLeaderChangeValidatingPositions(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testAssignAndConsume(String, String) > testAssignAndConsume(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testAutoCommitOnClose(String, String) > testAutoCommitOnClose(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testAutoCommitOnClose(String, String) > testAutoCommitOnClose(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testAutoCommitOnClose(String, String) > testAutoCommitOnClose(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testListTopics(String, String) > testListTopics(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testListTopics(String, String) > testListTopics(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testListTopics(String, String) > testListTopics(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testExpandingTopicSubscriptions(String, String) > testExpandingTopicSubscriptions(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testSubscribeAndCommitSync(String, String) > testSubscribeAndCommitSync(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMultiConsumerDefaultAssignor(String, String) > testMultiConsumerDefaultAssignor(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMultiConsumerDefaultAssignor(String, String) > testMultiConsumerDefaultAssignor(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMultiConsumerDefaultAssignor(String, String) > testMultiConsumerDefaultAssignor(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testInterceptors(String, String) > testInterceptors(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testInterceptors(String, String) > testInterceptors(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testInterceptors(String, String) > testInterceptors(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testConsumingWithEmptyGroupId(String, String) > testConsumingWithEmptyGroupId(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testConsumingWithEmptyGroupId(String, String) > testConsumingWithEmptyGroupId(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testConsumingWithEmptyGroupId(String, String) > testConsumingWithEmptyGroupId(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPatternUnsubscription(String, String) > testPatternUnsubscription(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPatternUnsubscription(String, String) > testPatternUnsubscription(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPatternUnsubscription(String, String) > testPatternUnsubscription(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testGroupConsumption(String, String) > testGroupConsumption(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPartitionsFor(String, String) > testPartitionsFor(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPartitionsFor(String, String) > testPartitionsFor(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPartitionsFor(String, String) > testPartitionsFor(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMultiConsumerDefaultAssignorAndVerifyAssignment(String, String) > testMultiConsumerDefaultAssignorAndVerifyAssignment(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMultiConsumerDefaultAssignorAndVerifyAssignment(String, String) > testMultiConsumerDefaultAssignorAndVerifyAssignment(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMultiConsumerDefaultAssignorAndVerifyAssignment(String, String) > testMultiConsumerDefaultAssignorAndVerifyAssignment(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testAutoCommitOnRebalance(String, String) > testAutoCommitOnRebalance(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testAutoCommitOnRebalance(String, String) > testAutoCommitOnRebalance(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testAutoCommitOnRebalance(String, String) > testAutoCommitOnRebalance(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testInterceptorsWithWrongKeyValue(String, String) > testInterceptorsWithWrongKeyValue(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testInterceptorsWithWrongKeyValue(String, String) > testInterceptorsWithWrongKeyValue(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testInterceptorsWithWrongKeyValue(String, String) > testInterceptorsWithWrongKeyValue(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPerPartitionLeadWithMaxPollRecords(String, String) > testPerPartitionLeadWithMaxPollRecords(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testHeaders(String, String) > testHeaders(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMaxPollIntervalMsDelayInAssignment(String, String) > testMaxPollIntervalMsDelayInAssignment(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMaxPollIntervalMsDelayInAssignment(String, String) > testMaxPollIntervalMsDelayInAssignment(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMaxPollIntervalMsDelayInAssignment(String, String) > testMaxPollIntervalMsDelayInAssignment(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testHeadersSerializerDeserializer(String, String) > testHeadersSerializerDeserializer(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testDeprecatedPollBlocksForAssignment(String, String) > testDeprecatedPollBlocksForAssignment(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testDeprecatedPollBlocksForAssignment(String, String) > testDeprecatedPollBlocksForAssignment(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testDeprecatedPollBlocksForAssignment(String, String) > testDeprecatedPollBlocksForAssignment(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPartitionPauseAndResume(String, String) > testPartitionPauseAndResume(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testQuotaMetricsNotCreatedIfNoQuotasConfigured(String, String) > testQuotaMetricsNotCreatedIfNoQuotasConfigured(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testRemoteAssignorInvalid(String, String) > testRemoteAssignorInvalid(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testAssignAndCommitSyncNotCommitted(String, String) > testAssignAndCommitSyncNotCommitted(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPerPartitionLagMetricsCleanUpWithSubscribe(String, String) > testPerPartitionLagMetricsCleanUpWithSubscribe(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPerPartitionLagMetricsCleanUpWithSubscribe(String, String) > testPerPartitionLagMetricsCleanUpWithSubscribe(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPerPartitionLagMetricsCleanUpWithSubscribe(String, String) > testPerPartitionLagMetricsCleanUpWithSubscribe(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testConsumeMessagesWithLogAppendTime(String, String) > testConsumeMessagesWithLogAppendTime(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPerPartitionLagMetricsWhenReadCommitted(String, String) > testPerPartitionLagMetricsWhenReadCommitted(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMultiConsumerRoundRobinAssignor(String, String) > testMultiConsumerRoundRobinAssignor(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMultiConsumerRoundRobinAssignor(String, String) > testMultiConsumerRoundRobinAssignor(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMultiConsumerRoundRobinAssignor(String, String) > testMultiConsumerRoundRobinAssignor(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testAutoCommitOnCloseAfterWakeup(String, String) > testAutoCommitOnCloseAfterWakeup(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testAutoCommitOnCloseAfterWakeup(String, String) > testAutoCommitOnCloseAfterWakeup(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testAutoCommitOnCloseAfterWakeup(String, String) > testAutoCommitOnCloseAfterWakeup(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testAssignAndCommitSyncAllConsumed(String, String) > testAssignAndCommitSyncAllConsumed(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testAssignAndFetchCommittedOffsets(String, String) > testAssignAndFetchCommittedOffsets(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testMaxPollRecords(String, String) > testMaxPollRecords(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testAutoOffsetReset(String, String) > testAutoOffsetReset(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPerPartitionLagWithMaxPollRecords(String, String) > testPerPartitionLagWithMaxPollRecords(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testFetchInvalidOffset(String, String) > testFetchInvalidOffset(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testAutoCommitIntercept(String, String) > testAutoCommitIntercept(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testAutoCommitIntercept(String, String) > testAutoCommitIntercept(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testAutoCommitIntercept(String, String) > testAutoCommitIntercept(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testFetchOutOfRangeOffsetResetConfigLatest(String, String) > testFetchOutOfRangeOffsetResetConfigLatest(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testFetchOutOfRangeOffsetResetConfigEarliest(String, String) > testFetchOutOfRangeOffsetResetConfigEarliest(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testStaticConsumerDetectsNewPartitionCreatedAfterRestart(String, String) > testStaticConsumerDetectsNewPartitionCreatedAfterRestart(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testStaticConsumerDetectsNewPartitionCreatedAfterRestart(String, String) > testStaticConsumerDetectsNewPartitionCreatedAfterRestart(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testStaticConsumerDetectsNewPartitionCreatedAfterRestart(String, String) > testStaticConsumerDetectsNewPartitionCreatedAfterRestart(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testFetchHonoursMaxPartitionFetchBytesIfLargeRecordNotFirst(String, String) > testFetchHonoursMaxPartitionFetchBytesIfLargeRecordNotFirst(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testCommitSpecifiedOffsets(String, String) > testCommitSpecifiedOffsets(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPerPartitionLeadMetricsCleanUpWithSubscribe(String, String) > testPerPartitionLeadMetricsCleanUpWithSubscribe(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPerPartitionLeadMetricsCleanUpWithSubscribe(String, String) > testPerPartitionLeadMetricsCleanUpWithSubscribe(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPerPartitionLeadMetricsCleanUpWithSubscribe(String, String) > testPerPartitionLeadMetricsCleanUpWithSubscribe(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testCommitMetadata(String, String) > testCommitMetadata(String, String).quorum=kraft+kip848.groupProtocol=consumer PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testRoundRobinAssignment(String, String) > testRoundRobinAssignment(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testRoundRobinAssignment(String, String) > testRoundRobinAssignment(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testRoundRobinAssignment(String, String) > testRoundRobinAssignment(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPatternSubscription(String, String) > testPatternSubscription(String, String).quorum=zk.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPatternSubscription(String, String) > testPatternSubscription(String, String).quorum=kraft.groupProtocol=generic PASSED
Gradle Test Run :core:integrationTest > Gradle Test Executor 3 > PlaintextConsumerTest > testPatternSubscription(String, String) > testPatternSubscription(String, String).quorum=kraft+kip848.groupProtocol=generic PASSED
```

lucasbru commented:
@philipnee thanks for the updates.
@cadonna I'll wait for you green light this by approving.

lianetm commented:
@philipnee thanks for the updates! One last comment. Important follow-up to this PR is to enable the specific max poll related integration tests (like `testMaxPollIntervalMs`), that are currently disabled because of the dependency on the callbacks. Is there a jira for that so we make sure we run them asap? Thanks!

philipnee committed:
clean up

rebase fix

philipnee commented:
Hey @lianetm - You are right. It uses rebalance callback in the test so it will need to be enabled after implementing the rebalance callback.  However, I did verify the leave group request by setting a break point.  Is there a jira for the rebalance callback invocation? I assume it should part of the task.

philipnee commented:
Per request: https://issues.apache.org/jira/browse/KAFKA-15993

philipnee committed:
Update HeartbeatRequestManager.java

philipnee committed:
increase max poll interval since it makes the test flaky

philipnee commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -948,6 +948,7 @@ class PlaintextConsumerTest extends BaseConsumerTest {
     // 1 consumer using range assignment
     this.consumerConfig.setProperty(ConsumerConfig.GROUP_ID_CONFIG, "range-group")
     this.consumerConfig.setProperty(ConsumerConfig.GROUP_REMOTE_ASSIGNOR_CONFIG, "range")
+    this.consumerConfig.setProperty(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG, "30000")
```
the test is pretty flaky with the default 6s so I set it to a higher value.

lucasbru commented:
Test failures are unrelated.

brandboat commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:
```diff
@@ -116,30 +122,33 @@ public HeartbeatRequestManager(
+        this.heartbeatState = new HeartbeatState(subscriptions, membershipManager, maxPollIntervalMs);
         this.heartbeatRequestState = new HeartbeatRequestState(logContext, time, 0, retryBackoffMs,
-            retryBackoffMaxMs, rebalanceTimeoutMs);
+            retryBackoffMaxMs, maxPollIntervalMs);
```
Hi all, and apologies for jumping in on an older PR. I noticed we use `max.poll.interval.ms` (default 300000) as the jitter value. Although `RequestState.remainingBackoffMs` guards with Math.max(0, …), which won’t make the value become negative, using a jitter that isn’t in (0–1) seems unexpected. Was there a design consideration behind this?

lucasbru replied:
This indeed looks like a bug! Nice catch.

chia7712 replied:
@brandboat please file a JIRA for this bug fix. Thanks for finding this

brandboat replied:
Here you go! https://issues.apache.org/jira/browse/KAFKA-19813
